gemrb/plugins/SDLAudio/SDLAudio.cpp

```diff
@@ -133,9 +143,9 @@ void SDLAudio::music_callback(void *udata, uint8_t *stream, int len)
 	int mixerLen = len;
 
 	SDLAudio *driver = (SDLAudio *)udata;
+	driver->MusicMutex.lock();
```

Comment:
why did you make this change? never yielding a lock in an infinite loop is a code smell at best.

Reply:
I was like that before, I can change it. But isn't there a chance of `MusicReader` change during that `while(true)` check?

Reply:
thats why `MusicReader` is _supposed_ to be protected by a mutex. I believe I noted this in my other comments.

Reply:
Changed. Still not quite sure why it's better to keep the guard inside the loop. Isn't it slower, since it's locking/unlocking each iteration? And what about possible change of `MusicReader` during condition check on every iteration? That could mess up things too, since we'd want to continue reading stuff from the same buffer.
If locking/unlocking is a bad thing to do maybe it's better to place guard and while loop inside additional `{ }` code block?

Reply:
its not about being _slow_ or _fast_, it is about being _correct_. The reason why you practically never should lock a mutex outside of an infinite loop is you've now locked every other thread out of reading/writing whatever that mutex is protecting. If you are looping that _implies_ you expect some condition to change, but don't know when (eg you expect another thread to do something to change this state to break from the loop). Almost certainly this thing you expect to change _requires_ the very mutex you locked and now you either have a race condition or a deadlock.

Your concern about `MusicReader` changing is exactly why you should have the `lock_guard`; `MusicReader` _shouldn't_ be read/written by anything not owning the mutex. The other thread will block waiting to make the change while this thread holds the mutex. then the mutex becomes unlocked at the bottom of the loop and the other thread can grab it then this thread blocks at the top of the loop waiting for the mutex (eg the `MusicReader` change) to become available.

Reply:
>then the mutex becomes unlocked at the bottom of the loop and the other thread can grab it then this thread blocks at the top of the loop waiting for the mutex (eg the MusicReader change) to become available.

`music_callback` doesn't rely on any other thread for switching MusicReader (since it can do this itself or break the loop if there's no songs left to play), but yeah, it shouldn't really matter if it's finishing reading stuff from the next song on the same playlist or from a different one. So guard lock should work here fine indeed.

Lock in the `buffer_callback` should still be outside the loop tho. I think that `FreeBuffers()` can mess it up pretty badly if it's run on the loop condition check.

Also fixed one more bug while looking through the code.
`stream = stream + cnt;`
cnt is number of short samples and stream is  uint_8*. This resulted in empty ending and some overlapping of the buffer on song changes.